network/blob_service.go

```diff
@@ -0,0 +1,179 @@
+	"github.com/ipfs/go-datastore"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	provider "github.com/ipfs/go-ipfs-provider"
+	"github.com/ipfs/go-ipfs-provider/simple"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"github.com/libp2p/go-libp2p-core/routing"
+
+	"github.com/onflow/flow-go/module/blobs"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+)
+
+// BlobGetter is the common interface shared between blobservice sessions and
+// the blobservice.
+type BlobGetter interface {
+	// GetBlob gets the requested blob.
+	GetBlob(ctx context.Context, c cid.Cid) (blobs.Blob, error)
+
+	// GetBlobs does a batch request for the given cids, returning blobs as
+	// they are found, in no particular order.
+	//
+	// It may not be able to find all requested blobs (or the context may
+	// be canceled). In that case, it will close the channel early. It is up
+	// to the consumer to detect this situation and keep track which blobs
+	// it has received and which it hasn't.
+	GetBlobs(ctx context.Context, ks []cid.Cid) <-chan blobs.Blob
+}
+
+// BlobService is a hybrid blob datastore. It stores data in a local
+// datastore and may retrieve data from a remote Exchange.
+// It uses an internal `datastore.Datastore` instance to store values.
+type BlobService interface {
+	component.Component
+	BlobGetter
+
+	// AddBlob puts a given blob to the underlying datastore
+	AddBlob(ctx context.Context, b blobs.Blob) error
+
+	// AddBlobs adds a slice of blobs at the same time using batching
+	// capabilities of the underlying datastore whenever possible.
+	AddBlobs(ctx context.Context, bs []blobs.Blob) error
+
+	// DeleteBlob deletes the given blob from the blobservice.
+	DeleteBlob(ctx context.Context, c cid.Cid) error
+
+	// GetSession creates a new session that allows for controlled exchange of wantlists to decrease the bandwidth overhead.
+	GetSession(ctx context.Context) BlobGetter
+
+	// TriggerReprovide updates the BlobService's provider entries in the DHT
+	TriggerReprovide(ctx context.Context) error
+}
+
+type blobService struct {
+	*component.ComponentManager
+	blockService blockservice.BlockService
+	reprovider   provider.Reprovider
+}
+
+var _ BlobService = (*blobService)(nil)
+var _ component.Component = (*blobService)(nil)
+
+type BlobServiceConfig struct {
+	ReprovideInterval time.Duration // the interval at which the DHT provider entries are refreshed
+}
+
+type BlobServiceOption func(*BlobServiceConfig)
+
+// WithReprovideInterval sets the interval at which DHT provider entries are refreshed
+func WithReprovideInterval(d time.Duration) BlobServiceOption {
+	return func(config *BlobServiceConfig) {
+		config.ReprovideInterval = d
+	}
+}
+
+// NewBlobService creates a new BlobService.
+func NewBlobService(
+	host host.Host,
+	r routing.ContentRouting,
+	prefix string,
+	ds datastore.Batching,
+	opts ...BlobServiceOption,
+) BlobService {
```

Comment:
From the POV of the Flow DPS, I am a bit concerned about those parameters that need to be provided. Since the Blob Service is a required component of the Execution Data Service, we need to instantiate it, and I'm not sure how we'll know  how to configure it, from what I see here.

Where do the following come from?

* Routing information
* Prefix
* Datastore Batching

How should we expect to configure this in the Flow DPS? Should it be configuration flags that are set by the user at launch, or parsed from a configuration file? Fetched from other nodes in the network?

Reply:
We will provide examples for all of those later.

ContentRouting in our specific case would be a DHT instance: https://pkg.go.dev/github.com/libp2p/go-libp2p-kad-dht#section-readme

Currently, the the libp2p host and DHT instance are both created internally by the ConsensusFollower, due to the way the code was already structured. However, we have an open issue to make both of these injectable: https://github.com/onflow/flow-go/issues/1669

Think of the Prefix as the protocol name. This is just some agreed upon string that all parties participating in the state sync protocol will use. The reason we support multiple prefixes is because the BlobService is not restricted to the single ExecutionDataService usecase, but it may be used as the underlying data transport protocol for various different high level protocols.

The datastore can be whatever implementation you like. We will be using a modified version of https://pkg.go.dev/github.com/ipfs/go-ds-badger.